### PR TITLE
 MISUV-6717: Balancing Payment : Display zero as applicable

### PR DESCRIPTION
**Pull request by SiddharthAtos**

### CODE REVIEW

File: app/views/partials/chargeSummary/ChargeSummaryNotCodingOutAndNotPayeSelfAssessment.scala.html [hunk 1]
```diff
@@ -43,8 +43,14 @@
             <span class="govuk-tag govuk-tag--red">@messages("chargeSummary.overdue")</span>
         }
         @{
-            if(latePaymentInterestCharge) documentDetail.interestEndDate.getOrElse(throw MissingFieldException("Interest End Date")).toLongDate
-            else documentDetail.documentDueDate.map(_.toLongDate).getOrElse(messages("chargeSummary.na"))
+            if(latePaymentInterestCharge) {
+                documentDetail.interestEndDate.getOrElse(throw MissingFieldException("Interest End Date")).toLongDate
```
Comment (avelx) [inline app/views/partials/chargeSummary/ChargeSummaryNotCodingOutAndNotPayeSelfAssessment.scala.html:47]: No exception to be throw in the views pls

File: app/controllers/TaxYearSummaryController.scala [hunk 2]
```diff
@@ -168,7 +168,20 @@ class TaxYearSummaryController @Inject()(taxYearSummaryView: TaxYearSummary,
...
+                  originalAmount = documentDetail.originalAmount,
+                  transactionId = documentDetail.transactionId,
+                  outstandingAmount = None,
+                  documentDate = documentDetail.documentDate), None)
+          }
+        }
+        f(documentDetailsWithDueDates ++ documentDetailsWithDueDatesForLpi ++ documentDetailsWithDueDatesCodingOutPaye ++ documentDetailsWithDueDatesCodingOut ++ documentDetailZeroBalancingCharge)
       case FinancialDetailsErrorModel(NOT_FOUND, _) => f(List.empty)
       case _ if isAgent =>
         Logger("application").error(s"[TaxYearSummaryController][withTaxYearFinancials] - Could not retrieve financial details for year: $taxYear")
```
Comment (EdwardDennis) [inline app/controllers/TaxYearSummaryController.scala:187]: This logging agent login logic is flipped as are other ones in this file, can you fix?

Comment (SiddharthAtos) [reply app/controllers/TaxYearSummaryController.scala:187]: ok

File: app/controllers/TaxYearSummaryController.scala [hunk 3]
```diff
@@ -168,7 +168,20 @@ class TaxYearSummaryController @Inject()(taxYearSummaryView: TaxYearSummary,
             documentDetail => DocumentDetailWithDueDate(documentDetail, documentDetail.getDueDate(),
               dunningLock = financialDetails.dunningLockExists(documentDetail.transactionId)))
         }
-        f(documentDetailsWithDueDates ++ documentDetailsWithDueDatesForLpi ++ documentDetailsWithDueDatesCodingOutPaye ++ documentDetailsWithDueDatesCodingOut)
+        val documentDetailZeroBalancingCharge: List[DocumentDetailWithDueDate] = {
+          documentDetails.filter(_.isBalancingCharge(isEnabled(CodingOut))).filter(_.originalAmount == Some(BigDecimal(0))).map {
```
Comment (EdwardDennis) [inline app/controllers/TaxYearSummaryController.scala:172]: Should fix: originalAmount filtering is model logic, moving to isBalancingCharge in DocumentDetail would ensure all relevant logic is in one place. See topmost methods in DocumentDetail for similar examples

File: app/views/ChargeSummary.scala.html [hunk 4]
```diff
@@ -158,6 +158,10 @@
     }
 }
 
+@isBalancingChargeZero = @{
+    val isBalancingCharge = documentDetail.isBalancingCharge(codingOutEnabled)
+    documentDetail.originalAmount.map( _ == BigDecimal(0) && isBalancingCharge ).getOrElse(false)
+}
```
Comment (EdwardDennis) [inline app/views/ChargeSummary.scala.html:164]: As above, this logic should be in the model

Comment (EdwardDennis) [reply app/views/ChargeSummary.scala.html:164]: You could use exists here to avoid getOrElse

File: app/views/TaxYearSummary.scala.html [hunk 5]
```diff
@@ -220,6 +223,14 @@ <h2 class="govuk-heading-m">@messages(header)</h2>
                     } else if(codingOutEnabled && documentDetailWithDueDate.documentDetail.isPayeSelfAssessment) {
                         <td class="govuk-table__cell">@messages("tax-year-summary.na")</td>
                         <td class="govuk-table__cell govuk-table__cell--numeric">@documentDetailWithDueDate.documentDetail.originalAmount.getOrElse(throw MissingFieldException("Original Amount")).toCurrencyString</td>
+                    }else if(documentDetailWithDueDate.documentDetail.isBalancingCharge(codingOutEnabled)) {
```
Comment (EdwardDennis) [inline app/views/TaxYearSummary.scala.html:226]: Should fix:
Missing space after curly brace

File: app/views/partials/chargeSummary/ChargeSummaryNotCodingOutAndNotPayeSelfAssessment.scala.html [hunk 6]
```diff
@@ -43,8 +43,14 @@
...
+            } else if(documentDetail.isBalancingCharge(codingOutEnabled)) {
+                val dueDate = documentDetail.documentDueDate.map(_.toLongDate).getOrElse(messages("tax-year-summary.na"))
+                documentDetail.originalAmount.filter( _ != 0 ).map( _ => dueDate).getOrElse(messages("tax-year-summary.na"))
```
Comment (EdwardDennis) [inline app/views/partials/chargeSummary/ChargeSummaryNotCodingOutAndNotPayeSelfAssessment.scala.html:50]: Should be moved to model

File: app/views/ChargeSummary.scala.html [hunk 7]
```diff
@@ -278,11 +282,11 @@ <h1 class="govuk-heading-xl">
         @paymentProcessingInfo
     }
 
-    @if(isAgent && documentDetail.remainingToPay > 0 && !(codingOutEnabled && documentDetail.isPayeSelfAssessment)){
+    @if(isAgent && documentDetail.remainingToPay > 0 && !(codingOutEnabled && documentDetail.isPayeSelfAssessment) && !isBalancingChargeZero){
```
Comment (joesheridan) [inline app/views/ChargeSummary.scala.html:285]: we don't need to show payment options in the case of a zero amount balancing charge (nothing to pay) - so existing logic should work without changes, no?

Comment (SiddharthAtos) [reply app/views/ChargeSummary.scala.html:285]: right. will make the change

File: app/models/financialDetails/DocumentDetail.scala [hunk 8]
```diff
@@ -158,6 +163,22 @@ case class DocumentDetail(taxYear: Int,
     case _ => false
   }
 
+  def isBalancingCharge(codedOutEnabled: Boolean = false): Boolean = getChargeTypeKey(codedOutEnabled) == "balancingCharge.text"
+
+  def isBalancingChargeZero(codedOutEnabled: Boolean = false): Boolean = {
+    (isBalancingCharge(codedOutEnabled), this.originalAmount) match {
+      case (true, Some(value)) => if (value == BigDecimal(0.0)) true else false
```
Comment (EdwardDennis) [inline app/models/financialDetails/DocumentDetail.scala:170]: Could simplify: no need for if/else, could evaluate value == BigDecimal(0.0) directly

Comment (SiddharthAtos) [reply app/models/financialDetails/DocumentDetail.scala:170]: right!! :)

File: app/views/partials/chargeSummary/ChargeSummaryNotCodingOutAndNotPayeSelfAssessment.scala.html (hunk 1 above)
Comment (SiddharthAtos) [reply app/views/partials/chargeSummary/ChargeSummaryNotCodingOutAndNotPayeSelfAssessment.scala.html:47]: There are multiple statements like that in that file. TechDebt would be better.

**Review by EdwardDennis — APPROVED** [review]

**Pull request merged**